### PR TITLE
Quit after command-line parses env var

### PR DESCRIPTION
**Pull request by tclose**
This PR addresses #2679 by defining an environment variable ("MRTRIX_PARSE_ONLY"), which if set to "1", will cause all commands to terminate after successful completion of the command-line parsing with a "warning" stating that the CL args have been parsed successfully.

@Lestropie @jdtournier

### CODE REVIEW

**Comment by tclose** [comment]
@Lestropie @jdtournier this PR is ready for review if you have a moment

**Comment by tclose** [comment]
NB: Just rebased it onto dev instead of master

**Review by Lestropie — CHANGES_REQUESTED** [review]
Switched PR target branch to match rebasing to `dev`.

1.  Currently early exit does not apply when running in an R environment. I don't know the extent to which that is used beyond the original implementer... nevertheless, given the definition and claimed implementation of the environment variable, it should probably be doing the same thing there also.
2.  Not sure if "parse only" is adequately descriptive. In the context of CLI / workflow engine, sure, but in the context of the software as a whole, maybe something more verbose like "CLI parse only"? @jdtournier?

**Comment by jdtournier** [comment]
> Currently early exit does not apply when running in an R environment. I don't know the extent to which that is used beyond the original implementer... 

Not at all used as far as I know... I don't think the original implementer is even still working in the field! I've been meaning to remove support for R integration, it always felt like a bit of a hack. I'll work on that when I have a minute. But in the meantime, I really wouldn't worry a great deal about this!

**Comment by Lestropie** [comment]
> Which commands are R based?

All C++ commands; see parallel functions in `core/command.h`.

**Comment by tclose** [comment]
> > Which commands are R based?
> 
> All C++ commands; see parallel functions in `core/command.h`.

Ah, I misunderstood. You mean all mrtrix commands can be called from R, rather than R-based mrtrix commands. If it is an issue, we could just add a disclaimer that it won't work when calling from R, but it sounds like it will be dropped in any case.

Is it good to merge now? Did you request a change somewhere in the code @Lestropie? Because I can't see where that is if so

File: core/command.h [hunk 1]
```diff
@@ -98,6 +99,18 @@ int main (int cmdline_argc, char** cmdline_argv)
     ::MR::GUI::App app (cmdline_argc, cmdline_argv);
 #endif
     ::MR::App::parse ();
+
+    //ENVVAR name: MRTRIX_CLI_PARSE_ONLY
+    //ENVVAR Set the command to parse the provided inputs and then quit
+    //ENVVAR if it is 1. This can be used in the CI of wrapping code,
+    //ENVVAR such as the automatically generated Pydra interfaces
+    char* parse_only = std::getenv("MRTRIX_CLI_PARSE_ONLY");
+    if (parse_only && ::MR::to<int>(parse_only)) {
```
Comment (Lestropie) [inline core/command.h:108]:
```suggestion
    if (parse_only && ::MR::to<bool>(parse_only)) {
```

https://github.com/MRtrix3/mrtrix3/blob/3.0.4/core/mrtrix.h#L285-L293

File: core/command.h [hunk 2]
```diff
@@ -98,6 +99,18 @@ int main (int cmdline_argc, char** cmdline_argv)
...
+      WARN(
+        "Quitting after parsing command-line arguments successfully because environment variable 'MRTRIX_CLI_PARSE_ONLY' is set to '1'"
+      );
```
Comment (Lestropie) [inline core/command.h:111]:
```suggestion
      INFO(
        "Quitting after parsing command-line arguments successfully due to environment variable 'MRTRIX_CLI_PARSE_ONLY'"
      );
```

-   Warning-level is for when there is some kind of ambiguity or uncertainty regarding how to proceed, that the user needs to be alerted to, but the command is going to attempt to proceed nevertheless.

-   Environment variable value shouldn't necessarily need to be "`1`".

File: lib/mrtrix3/app.py [hunk 3]
```diff
@@ -192,6 +192,15 @@ def _execute(module): #pylint: disable=unused-variable
   CMDLINE.print_citation_warning()
 
   return_code = 0
+
+  cli_parse_only = os.getenv("MRTRIX_CLI_PARSE_ONLY")
+  if cli_parse_only == "1":
+    warn(
+      "Quitting after parsing command-line arguments successfully because "
+      "environment variable 'MRTRIX_CLI_PARSE_ONLY' is set to '1'"
+    )
```
Comment (Lestropie) [inline lib/mrtrix3/app.py:201]:
```suggestion
  cli_parse_only = os.getenv('MRTRIX_CLI_PARSE_ONLY')
  if cli_parse_only:
    try:
      if cli_parse_only.lower() in ['yes', 'true'] or int(cli_parse_only):
        info(
          'Quitting after parsing command-line arguments successfully due to '
          'environment variable "MRTRIX_CLI_PARSE_ONLY"'
        )
        sys.exit(return_code)
    except ValueError:
      warn('Potentially corrupt environment variable "MRTRIX_CLI_PARSE_ONLY" '
           '= " + cli_parse_only + '; ignoring')
```

Needs testing.

-   The environment variable should be interpreted as a boolean value (ideally the *MRtrix3* interpretation of strings as bools should be functionalised somewhere for Python as it is for C++).

-   Should be INFO level rather than WARN.

File: pydra/src/pydra/tasks/mrtrix3/_version.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+# file generated by setuptools_scm
+# don't change, don't track in version control
+__version__ = version = '3.0.5.dev589+g9e2806609.d20230817'
+__version_tuple__ = version_tuple = (3, 0, 5, 'dev589', 'g9e2806609.d20230817')
```
Comment (Lestropie) [inline pydra/src/pydra/tasks/mrtrix3/_version.py:4]: Are these files intended to be included in the PR? Certainly the header comment suggests not.

File: pydra/src/pydra/tasks/mrtrix3/v3_0/fivetissuetype2gmwmi.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
+import typing as ty
```
Comment (Lestropie) [inline pydra/src/pydra/tasks/mrtrix3/v3_0/fivetissuetype2gmwmi.py:1]: This file also looks like an accidental addition not relevant to this specific PR.

**Comment by Lestropie** [comment]
Given the absence of interest in the R interface, perhaps just clarify very briefly in the description for the new environment variable that it only applies to the shell interface and not to R?

I've left a bunch of comments in the code, and you will need to resolve conflicts with latest `dev`; I'll take another look after that.

**Comment by tclose** [comment]
@Lestropie Ok, I have made those changes now so I believe it should be good to re-review. Note that there is a difference in behaviour in the message printing between the c++ and python, i.e. it shows up by default in python commands but only when the -info flag is passed to c++ commands, but this seems to be the case with other "info" messages as well.

(NB: I can't see the "1 change requested", do you know what that is referring to)

**Comment by Lestropie** [comment]
> it shows up by default in python commands but only when the -info flag is passed to c++ commands, but this seems to be the case with other "info" messages as well.

Uh, I suggested calling function `info()` [here](https://github.com/MRtrix3/mrtrix3/pull/2694/files/d5f3dd8e140d43493f31304ba0c3bca0bb5caa4e#r1305093728) rather than `warn()`, thinking it was an analogue to C++ `INFO()` macro, but function `app.info()` doesn't actually exist. So I'm honestly not sure what that is even calling...

My intention was for nothing to be printed to `stderr` unless `-info` was specified. I think that if the environment variable has been set, the command is being run in a setting where they don't need to read a user-friendly message to know that parsing completed successfully and the command is exiting prematurely. There's an argument to be made that perhaps this message should be `CONSOLE()` / `app.console()`-level in case a user somehow has this envvar set unexpectedly and needs to be notified of such. And that would also make it easier to get consistent behaviour between the two languages. I'm not feeling terribly strongly either way.

> I can't see the "1 change requested", do you know what that is referring to

[This comment](https://github.com/MRtrix3/mrtrix3/pull/2694#pullrequestreview-1585989384) was made as a review. It'll resolve once I re-review.

File: lib/mrtrix3/app.py [hunk 6]
```diff
@@ -192,6 +192,21 @@ def _execute(module): #pylint: disable=unused-variable
...
+    try:
+      if cli_parse_only.lower() in ['yes', 'true'] or int(cli_parse_only):
+        console(
+          'Quitting after parsing command-line arguments successfully due to '
+          'environment variable "MRTRIX_CLI_PARSE_ONLY"'
+        )
+        sys.exit(return_code)
+    except ValueError:
+      warn('Potentially corrupt environment variable "MRTRIX_CLI_PARSE_ONLY" '
+           '= "' + cli_parse_only + '; ignoring')
```
Comment (Lestropie) [inline lib/mrtrix3/app.py:207]:
```suggestion
           '= "' + cli_parse_only + '"; ignoring')
```

My bad.

File: core/command.h [hunk 7]
```diff
@@ -92,6 +93,18 @@ int main(int cmdline_argc, char **cmdline_argv) {
...
+
+    // ENVVAR name: MRTRIX_CLI_PARSE_ONLY
+    // ENVVAR Set the command to parse the provided inputs and then quit
+    // ENVVAR if it is set. This can be used in the CI of wrapping code,
+    // ENVVAR such as the automatically generated Pydra interfaces.
+    // ENVVAR Note that it will have no effect for R interfaces
+    char *parse_only = std::getenv("MRTRIX_CLI_PARSE_ONLY");
+    if (parse_only && ::MR::to<bool>(parse_only)) {
+      INFO("Quitting after parsing command-line arguments successfully due to environment variable "
+           "'MRTRIX_CLI_PARSE_ONLY'");
```
Comment (Lestropie) [inline core/command.h:105]:
```suggestion
      CONSOLE("Quitting after parsing command-line arguments successfully due to environment variable "
              "'MRTRIX_CLI_PARSE_ONLY'");
```

**Comment by tclose** [comment]
> > it shows up by default in python commands but only when the -info flag is passed to c++ commands, but this seems to be the case with other "info" messages as well.
> 
> Uh, I suggested calling function `info()` [here](https://github.com/MRtrix3/mrtrix3/pull/2694/files/d5f3dd8e140d43493f31304ba0c3bca0bb5caa4e#r1305093728) rather than `warn()`, thinking it was an analogue to C++ `INFO()` macro, but function `app.info()` doesn't actually exist. So I'm honestly not sure what that is even calling...
> 
> My intention was for nothing to be printed to `stderr` unless `-info` was specified. I think that if the environment variable has been set, the command is being run in a setting where they don't need to read a user-friendly message to know that parsing completed successfully and the command is exiting prematurely.

Maybe the code could be refactored to use Python's std logger, which would give you the behaviour you are after, but that seems like a fair bit of work for not much gain.

> There's an argument to be made that perhaps this message should be `CONSOLE()` / `app.console()`-level in case a user somehow has this envvar set unexpectedly and needs to be notified of such. And that would also make it easier to get consistent behaviour between the two languages. I'm not feeling terribly strongly either way.

This was my thinking with the warning, but in either case it probably isn't worth worrying about too much

> > I can't see the "1 change requested", do you know what that is referring to
> 
> [This comment](https://github.com/MRtrix3/mrtrix3/pull/2694#pullrequestreview-1585989384) was made as a review. It'll resolve once I re-review.

**Review by Lestropie — APPROVED** [review]
Only thing that still catches my eye is that if the contents of the envvar can't be converted to bool, the exception won't be handled neatly in C++. But realistically it's not an issue, this is a feature purely for developers anyway. So should be good to go from here.
